Introduction to Free Monads/article.md

```diff
@@ -0,0 +1,702 @@
+We'll pretend we "forgot" that it's a monad, and try to implement it from scratch in terms of a free monad.
+
+First, we define a typical `State` newtype, and let GHC auto-derive the functor instance:
+
+```haskell
+newtype StateF s a = StateF { runStateF :: s -> (a, s) }
+  deriving stock Functor
+```
+
+(note that we're calling it `StateF` for "state functor")
+
+We can also define the typical functions for working with state:
+
+```haskell
+getF :: StateF s s
+getF = StateF $ \s -> (s, s)
+
+putF :: s -> StateF s ()
+putF s = StateF $ \_ -> ((), s)
+```
+
+Now we can define a "free state" monad:
+
+```haskell
+type State s = Free (StateF s)
+```
+
+We'll also need to lift our state operations into the monad.
+We could do it manually, of course.
+We have `StateF s a`, and we need to get `StateF s (Free (StateF s) a)`, and wrap it in `Free`.
+To lift any `a` into `Free f a` we just need to apply `Pure`:
+
+```haskell
+get :: State s s
+get = Free $ Pure <$> getF
+```
+
+But since it is a common pattern, the `free` library provides a function to do this for us:
+
+```haskell
+liftF :: (Functor f, MonadFree f m) => f a -> m a
+```
+
+`MonadFree` is just an MTL-style type class for different encodings of the free monad.
+In our case, we can pretend that
+
+```haskell
+liftF :: Functor f => f a -> Free f a
+```
+
+Hence, we can define
+
+```haskell
+get :: State s s
+get = liftF get
+
+set :: s -> State s ()
+set = liftF . set
+```
+
+and, like magic, we have a `State` monad:
+
+```haskell
+someComputation :: State Int ()
+someComputation = do
+  i <- get
+  put $ i + 1
+  pure ()
+```
+
+There is a catch, however.
+This code will happily compile, but it doesn't do anything interesting.
+Indeed, we didn't define the meaning of the computation, only its form, and we can't directly use `runStateF` to "run" our free state monad.
+
+This means we need to write an interpreter, so let's do just that.
+Remember that `Free` has two constructors, `Pure` and `Free`, so we need to pattern-match on those:
+
+```haskell
+runState :: State s a -> s -> (a, s)
+runState (Pure x) s = (x, s)
+runState (Free f) s =
+  let (m, s') = runStateF f s
+  in runState m s'
+```
+
+If you're familiar with the usual state monad, you may have noticed that we've essentially moved the implementation of `>>=` to `runState`.
+The free monad doesn't specify the meaning of monadic actions, so we have to decide what those actions mean when we're running it.
+
+To illustrate this point, we're going to write another interpreter, which is essentially a pretty-printer:
+
+```haskell
+printState :: (Show s, Show a) => State s a -> s -> String
+printState (Pure x) s = "pure (" <> show x <> "," <> show s <> ")"
+printState (Free m) s =
+  let (x, s') = runStateF m s
+  in "state change " <> show s <> " -> "<> show s' <> "\n"
+    <> printState x s'
+```
+
+If we run `someComputateion` defined above through this interpreter, we'll get the following output:
+
+```
+state change 1 -> 1
```

Comment:
BTW, where is it defined that the initial state value is 1?

Reply:
Oh, indeed, it isn't. "Free state" is exactly the same as regular "state", behaviour-wise, so the initial state is supplied to `runState` (or in this case `printState`). But I've neglected to mention the exact call I'm doing to produce the output, which is of course `printState someComputation 1`.